src/main/java/com/algorand/algosdk/transaction/BoxReference.java

```diff
@@ -0,0 +1,105 @@
+package com.algorand.algosdk.transaction;
+
+import com.fasterxml.jackson.annotation.JsonCreator;
+import com.fasterxml.jackson.annotation.JsonInclude;
+import com.fasterxml.jackson.annotation.JsonProperty;
+import com.fasterxml.jackson.annotation.JsonPropertyOrder;
+
+import java.io.Serializable;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Objects;
+
+public class BoxReference {
+    // the appID of the app this box belongs to. Instead of serializing this value,
+    // it's used to calculate the appIdx for BoxReferenceSerialize.
+    public Long appID;
+
+    // the name of the box unique to the app it belongs to
+    public byte[] name;
+
+    public BoxReference(Long appID, byte[] name) {
+        this.appID = appID;
+        this.name = name;
+    }
+
+    @Override
+    public int hashCode() {
+        int result = Objects.hash(appID);
+        result = 31 * result + Arrays.hashCode(name);
+        return result;
+    }
+
+    @Override
+    public boolean equals(Object o) {
+        if (this == o) return true;
+        if (o == null || getClass() != o.getClass()) return false;
+        BoxReference that = (BoxReference) o;
+        return appID.equals(that.appID) &&
+                Arrays.equals(this.name, name);
+    }
+
+    // Foreign apps start from index 1.  Index 0 is the called App ID.
+    // Must apply offset to yield the foreign app index expected by algod.
+    private static final int FOREIGN_APPS_INDEX_OFFSET = 1;
+
+    public BoxReferenceSerialize getBoxReferenceSerialize(List<Long> foreignApps, Long curApp) {
+        if (appID == 0) {
+            return new BoxReferenceSerialize(0, name);
+        }
+
+        int appIdx = -1;
+        if (appID == curApp) {
+            appIdx = 0;
+        }
+
+        if (foreignApps != null) {
+            for (int i = 0; i < foreignApps.size(); i++) {
```

Comment:
Could we use `indexOf` here, e.g. `foreignApps.indexOf(appID)`, then add 1 to  `appIdx` if found (or always add 1 and see if `appIdx` equals 0

Reply:
@algochoi The feedback makes sense - Here's an attempt to implement the suggestion:  https://github.com/algorand/java-algorand-sdk/pull/340/commits/d225cde0bddc3441cc7bb5fd682f573fc01c8b44.